### PR TITLE
Fixes impurities on foam, smoke and cigs. Improves how impurities are handled.

### PR DESCRIPTION
**Pull request by Thalpy**
## About The Pull Request

Reworks the main handlier for impurities and adds it to the add_reagent proc instead. Adds an option to bypass the purity splitting too. This reduces the amount of add_reagent calls and will prevent any unintended on_mob_add() and delete calls for reagents that are removed from splitting.
Fixes smoke, foam and cigarette issues too - since they used copy_to isntead of trans_to.

## Why It's Good For The Game

Fixes bugs and unintended proc calls while improving performance.

## Changelog
:cl:
fix: Fixes impurities on foam, smoke and cigs.
code: Improved how impurities are handled
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
what was the previous (incorrect) effect that this PR is fixing?

**Comment by Thalpy** [comment]
> what was the previous (incorrect) effect that this PR is fixing?

Smoke and foam wouldn't split on mobs, cigarettes would spam runtimes and on mob add/delete would be called for the base reagent that was being inverted.

**Comment by LemonInTheDark** [comment]
Conflict du merge

**Comment by Thalpy** [comment]
Should be fixed now! Lets just make sure the checks pass.

**Comment by LemonInTheDark** [comment]
Spurious unit test failure, get the flyswatter

**Comment by Thalpy** [comment]
It seems to have disappeared, was it from another pr?

**Comment by LemonInTheDark** [comment]
Ye

**Comment by Thalpy** [comment]
It should be okay to merge now, right?

**Comment by LemonInTheDark** [comment]
Need to actually read the code first, but sure

File: code/modules/reagents/chemistry/holder.dm [hunk 1]
```diff
@@ -167,13 +168,19 @@
...
-		added_purity = D.creation_purity //Usually 1
-
+		added_purity = glob_reagent.creation_purity //Usually 1
 	if(!added_ph)
-		added_ph = D.ph
+		added_ph = glob_reagent.ph
+
+	//Split up the reagent if it's in a mob
+	var/has_split = FALSE
+	if(is_type_in_list(my_atom, list(/mob/living/carbon, /obj/item/organ/stomach)) && !ignore_splitting) //Stomachs are a pain - they will constantly call on_mob_add unless we split on addition to stomachs, but we also want to make sure we don't double split
```
Comment (LemonInTheDark) [inline code/modules/reagents/chemistry/holder.dm:178]: Is this the only way to do this? Cause I really don't like this

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemistry/holder.dm (hunk 1 above)
Comment (Thalpy) [reply code/modules/reagents/chemistry/holder.dm:178]: I could make it a holder flag?

Comment (Thalpy) [reply code/modules/reagents/chemistry/holder.dm:178]: I'll make it a holder flag

File: code/__DEFINES/reagents.dm [hunk 2]
```diff
@@ -13,6 +13,8 @@
 #define AMOUNT_VISIBLE (1<<6) // For non-transparent containers that still have the general amount of reagents in them visible.
 #define NO_REACT (1<<7) // Applied to a reagent holder, the contents will not react with each other.
 #define REAGENT_HOLDER_INSTANT_REACT (1<<8)  // Applied to a reagent holder, all of the reactions in the reagents datum will be instant. Meant to be used for things like smoke effects where reactions aren't meant to occur
+///If the holder is "alive" (i.e. mobs and organs)
+#define REAGENT_HOLDER_ALIVE (1<<9)
```
Comment (LemonInTheDark) [inline code/__DEFINES/reagents.dm:18]: Do you mind documenting why it exists?

Comment (Thalpy) [reply code/__DEFINES/reagents.dm:18]: Sure!

**Pull request merged**